### PR TITLE
#162949209 Add descriptive validation error messages

### PR DESCRIPTION
**Pull request by Allan690**
#### What does this PR do?
Adds descriptive error messages for the registration endpoint
#### Description of Task to be completed?
- Implement error messages for the registration endpoint
#### How should this be manually tested?
1. Clone the repo: `git clone https://github.com/andela/ah-the-unsullied.git`
2. Make cloned directory the active directory: `cd ah-the-unsullied`
3. Checkout to the current branch: `git checkout ft-descriptive-validation-errors-162949211`
4. Set up a virtual environment and activate it.
```
- virtualenv <name_of_env>
- source <name_of_env>/bin/activate
```
5. Create a  .env file by doing `touch .env` in the current directory(root) and enter the database connection information in the format shown below: 
```
export DATABASE_URL=postgres://USERNAME:PASSWORD@HOST:5432/NAME
```
6. Create the database on terminal: `createdb ah_unsullied`
7.  Export the environment variables: `source .env`
8. Make migrations: `python manage.py makemigrations && python manage.py migrate`
9. Create a super user: `python manage.py createsuperuser`
10. Run the server: `python manage.py runserver`
11. Run the tests: `python manage.py test`
12. On Postman, navigate to [http://127.0.0.1/api/user/](http://127.0.0.1/api/user/) and enter the payload in this format:
```
{
"user": 
         {
     "username": "your_username",
      "email": "your_email",
      "password": "your_password"
          }
}
```

#### Any background context you want to provide?
None
#### What are the relevant pivotal tracker stories?
[#162949211](https://www.pivotaltracker.com/story/show/162949211)
#### Screenshots (if appropriate)
(some of the use cases handled)
- When user enters an existing username/email
screenshot 2019-01-11 at 18 10 31

- When the user enters invalid details
screenshot 2019-01-11 at 18 12 21

### CODE REVIEW

**Review by MbugwaSami — DISMISSED** [review]

**Review by kelvinwm — DISMISSED** [review]

**Review by nesh-dev — DISMISSED** [review]

**Review by MbugwaSami — APPROVED** [review]

**Review by MbugwaSami — DISMISSED** [review]

**Review by nesh-dev — DISMISSED** [review]

**Review by MbugwaSami — COMMENTED** [review]
LGTM

**Review by MbugwaSami — DISMISSED** [review]

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -9,19 +10,71 @@
...
+        error_messages={
+            'invalid': 'Email entered is invalid. Please enter a valid email address'
+        }
+    )
+
     # Ensure passwords are at least 8 characters long, no longer than 128
     # characters, and can not be read by the client.
-    password = serializers.CharField(
+    password = serializers.RegexField(
+        regex=r'^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[\!\@#\$%\^&]).*',
```
Comment (indungu) [inline authors/apps/authentication/serializers.py]: A strict password policy is good but this appears to be a little too strict. Reduce the restriction to a require a number and a special character, that should be good enough.

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -9,19 +10,71 @@
...
         max_length=128,
         min_length=8,
         write_only=True
     )
+    write_only = True,
+    error_messages = {
+        'max_length': 'Password allows a maximum of 128 characters.',
+        'min_length': 'Password allows a minimum of 8 characters.',
+        'invalid': 'Password must contain at least 1 uppercase letter, '
```
Comment (indungu) [inline authors/apps/authentication/serializers.py]: On making the change requested on line 56, please adjust the validation message appropriately.

File: authors/apps/authentication/serializers.py [hunk 3]
```diff
@@ -9,19 +10,71 @@
...
+
+    # Ensure the email entered is unique and has a descriptive error message
+    # when a duplicate email is entered and an invalid email address.
+    email = serializers.EmailField(
+        validators=[UniqueValidator(
+            queryset=User.objects.all(),
+            message='Email already exists. Please enter another email or sign in'
+        )],
+        error_messages={
+            'invalid': 'Email entered is invalid. Please enter a valid email address'
```
Comment (indungu) [inline authors/apps/authentication/serializers.py]: Avoid repetition. `Please enter a valid email.` is descriptive enough.

**Review by indungu — CHANGES_REQUESTED** [review]
Please implement the requested changes. Also be sure to update the screenshots appropriately.

**Review by MbugwaSami — APPROVED** [review]

**Review by sylviawanjiku — APPROVED** [review]

**Review by indungu — APPROVED** [review]

**Pull request merged**